### PR TITLE
check: Add support for Consul field tls_server_name

### LINKED ISSUE #2166: Service registration with fqdn rather than ip address
While registering services with nomad, service address is always set to ip of the service.

It would be very handy if there is an option to register services with fully qualified domain names especially when there is a proxy in the environment. 

Do you consider adding such an option in the near future? 

### Nomad version
Output from `nomad version`
Nomad v0.5.2-rc1

### PR DESCRIPTION
**Pull request by beautifulentropy**
### Description

The `tls_server_name` field was added to Consul in hashicorp/consul#9475. This PR adds support for the same field in Nomad `service -> check` stanzas.

### Motivation

There are two common situations where specifying this field can be beneficial:

1. When the check address is an IP, `tls_server_name` can be specified for SNI. Note: setting `tls_server_name` will also override the hostname used to verify the certificate presented by the server being checked.

2. When the hostname in the check address won't be present in the SAN (Subject Alternative Name) field of the certificate presented by the server being checked. Note: setting `tls_server_name` will also override the hostname used for SNI.

The latter of these is especially important given that today you would have to specify `tls_skip_verify` which omits TLS verification altogether, a choice which is generally unsuitable for production.

### Changes

- Add support for Consul health check field `tls_server_name` to `service -> check` stanzas.
- Add documentation to support this field.
- Fix up documentation for:
  - `tls_skip_verify` to indicate support for `https` and `grpc` with `grpc_use_tls` checks and
  - `grpc_use_tls` to indicate that `tls_server_name` can be combined with this field.

### Links

- [Related Consul PR](https://github.com/hashicorp/consul/pull/17481): includes an explanation of how this field is implemented inside of Consul, offers a slight tweak to what Consul should do when it's left unspecified, and updates their documentation similar to the docs updates in this PR.

Fixes #2166

### CODE REVIEW

File: nomad/structs/services.go [hunk 1]
```diff
@@ -465,6 +470,9 @@ func (sc *ServiceCheck) Hash(serviceID string) string {
 	// use name "true" to maintain ID stability
 	hashBool(h, sc.TLSSkipVerify, "true")
 
+	// Only include TLSServerName if set.
+	hashStringIfNonEmpty(h, sc.TLSServerName)
```
Comment (beautifulentropy) [inline nomad/structs/services.go:474]: Honestly not sure what the comment for this line should be. Let me know if this needs to change.

**Comment by beautifulentropy** [comment]
Note sure what, if anything, I can do about the `ember` test failure:

```shell
Run mainmatter/ember-asset-size-action@255fe534db3d2f731e0d70ce9a5de74b0a6b140a
No package-lock.json or yarn.lock detected! We strongly recommend committing one
/opt/hostedtoolcache/node/14.21.3/x64/bin/npm install
npm WARN saveError ENOENT: no such file or directory, open '/home/runner/work/nomad/nomad/package.json'
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN enoent ENOENT: no such file or directory, open '/home/runner/work/nomad/nomad/package.json'
npm WARN nomad No description
npm WARN nomad No repository field.
npm WARN nomad No README data
npm WARN nomad No license field.

up to date in 0.26[9](https://github.com/hashicorp/nomad/actions/runs/5094531551/jobs/9158463528?pr=17334#step:4:10)s
found 0 vulnerabilities

/opt/hostedtoolcache/node/[14](https://github.com/hashicorp/nomad/actions/runs/5094531551/jobs/9158463528?pr=17334#step:4:15).[21](https://github.com/hashicorp/nomad/actions/runs/5094531551/jobs/9158463528?pr=17334#step:4:22).3/x64/bin/npx ember build -prod
command not found: ember
Error: The process '/opt/hostedtoolcache/node/14.21.3/x64/bin/npx' failed with exit code 1
```

**Review by tgross — COMMENTED** [review]
Hi @beautifulentropy! Thanks for the PR! This looks great -- I've left a couple of minor comments. Can you run `make cl` as well to add a changelog entry?

I'm not sure what's wrong with the ember assets tests. I know we did some shuffling around of the GitHub Actions for that recently... might be worth rebasing if you branched from main more than a few days before you opened this PR. I might drag @philrenaud in here to see if that's something he has any idea what's up. Not something for you to worry about otherwise.

File: nomad/structs/services.go [hunk 2]
```diff
@@ -68,7 +68,8 @@ type ServiceCheck struct {
 	Interval               time.Duration       // Interval of the check
 	Timeout                time.Duration       // Timeout of the response from the check before consul fails the check
 	InitialStatus          string              // Initial status of the check
-	TLSSkipVerify          bool                // Skip TLS verification when Protocol=https
+	TLSServerName          string              // ServerName to use for SNI and TLS verification when (Type=https and Protocol=https) or (Type=grpc and GRPCUseTLS=true)
```
[the remark this answers is not in the document]
Comment (beautifulentropy) [reply nomad/structs/services.go:71]: I'm not sure if you're asking for a change in the first part of your comment but I've addressed the latter portion by adding validation in `validateNomad()`.

File: nomad/structs/services.go [hunk 3]
```diff
@@ -465,6 +470,9 @@ func (sc *ServiceCheck) Hash(serviceID string) string {
 	// use name "true" to maintain ID stability
 	hashBool(h, sc.TLSSkipVerify, "true")
 
+	// Only include TLSServerName if set.
```
[the remark this answers is not in the document]
Comment (beautifulentropy) [reply nomad/structs/services.go]: Thank you! I left a comment concerning this as I was unsure of what I should put here:

https://github.com/hashicorp/nomad/pull/17334#discussion_r1207289893

**Comment by beautifulentropy** [comment]
Not exactly sure what to do with these integration test failures. I can't repro most of them locally and some of them fail locally even on main.

File: nomad/structs/services.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tgross) [reply nomad/structs/services.go:71]: Sorry, that wasn't as clear as it could have been. Yes, you've got it right :+1:

**Review by tgross — APPROVED** [review]

**Comment by tgross** [comment]
Ah, so buried in the `test-groups (quick)` below lots of blank pages in the GHA interface was a genuine test failure in `nomad/structs/diff_test.go`. I've pushed up a quick fix for that test.

**Comment by tgross** [comment]
Ok, done! This will ship in the upcoming Nomad 1.6.0. Thanks @beautifulentropy!

**Pull request merged**